File: 1-30/twoSum.py

def twoSum(nums, target):
    number_to_index = {}
    print('len', len(nums))
    for i in range(len(nums)):
        print('i', i, 'nums[i]', nums[i])
        num_needed = target - nums[i]
        print('num_needed in number_to_index', num_needed in number_to_index, 'number_to_index', number_to_index, num_needed)
        if num_needed in number_to_index:
            return [number_to_index[num_needed], i]
        
        number_to_index[nums[i]] = i
    return None

nums = [8, 8, 5, 6]
target = 11
print(twoSum(nums, target))

File: 1-30/twoSum.java
import java.util.Map;
import java.util.HashMap;

public class twoSum {
    public static int[] twoSum(int[] nums, int target) {
        Map<Integer, Integer> numberToIndex = new HashMap<>();

        for (int i = 0; i < nums.length; i++) {
            int numNeeded = target - nums[i];
            
            System.out.println("numNeeded: " + numNeeded + " " + "numberToIndex.containsKey(numNeeded)" + " " + numberToIndex.containsKey(numNeeded));
            if (numberToIndex.containsKey(numNeeded)) {
                System.out.println("numberToIndex.get(numNeeded)" + numberToIndex.get(numNeeded));
                return new int[] {numberToIndex.get(numNeeded), i};
            }
            numberToIndex.put(nums[i], i);
        }

        return null;

    }


    public static void main(String[] args) {
        int[] nums = {8, 9, 6, 5, 7};
        int target = 12;

        int[] result = twoSum(nums, target);

       System.out.println("Indices: " + result[0] + ", " + result[1]);

    }
}